shared/entity/type.go

```diff
@@ -0,0 +1,332 @@
+func (t Type) Validate(allowEmpty bool) error {
+	if t == "" && allowEmpty {
+		return nil
+	}
+
+	if !shared.ValueInSlice(t, entityTypes) {
+		return fmt.Errorf("Unknown entity type %q", t)
+	}
+
+	return nil
+}
+
+// requiresProject returns true if an entity of the Type can only exist within the context of a project. Operations and
+// warnings may still be project specific but it is not an absolute requirement.
+func (t Type) requiresProject() (bool, error) {
+	err := t.Validate(false)
+	if err != nil {
+		return false, err
+	}
+
+	return !shared.ValueInSlice(t, []Type{TypeProject, TypeCertificate, TypeNode, TypeOperation, TypeStoragePool, TypeWarning, TypeClusterGroup}), nil
+}
+
+// nRequiredPathArguments returns the number of path arguments (mux variables) that are required to create a unique URL
+// for the given Type.
+func (t Type) nRequiredPathArguments() (int, error) {
+	err := t.Validate(false)
+	if err != nil {
+		return 0, err
+	}
+
+	typePath, err := t.path()
+	if err != nil {
+		return 0, err
+	}
+
+	nRequiredPathArguments := 0
+	for _, element := range typePath {
+		if element == pathPlaceholder {
+			nRequiredPathArguments++
+		}
+	}
+
+	return nRequiredPathArguments, nil
+}
+
+// URL returns a string URL for the Type.
+//
+// If the Type is project specific and no project name is given, the project name will be set to api.ProjectDefaultName.
+//
+// Warning: All arguments to this function will be URL encoded. They must not be URL encoded before calling this method.
+func (t Type) URL(projectName string, location string, pathArguments ...string) (Reference, error) {
+	requiresProject, err := t.requiresProject()
+	if err != nil {
+		return "", fmt.Errorf("Failed to check if entity type %q is project specific: %w", t, err)
+	}
+
+	if requiresProject && projectName == "" {
+		projectName = api.ProjectDefaultName
+	}
+
+	nRequiredPathArguments, err := t.nRequiredPathArguments()
+	if err != nil {
+		return "", fmt.Errorf("Failed to check number of required path arguments for entity type %q: %w", t, err)
+	}
+
+	if len(pathArguments) != nRequiredPathArguments {
+		return "", fmt.Errorf("Entity type %q requires `%d` path arguments but `%d` were given", t, nRequiredPathArguments, len(pathArguments))
+	}
+
+	pathParts, err := t.path()
+	if err != nil {
+		return "", fmt.Errorf("Failed to get path template for entity type %q: %w", t, err)
+	}
+
+	argIdx := 0
+	path := []string{version.APIVersion}
+	for _, pathPart := range pathParts {
+		if pathPart == pathPlaceholder {
+			pathPart = pathArguments[argIdx]
+			argIdx++
+		}
+
+		path = append(path, pathPart)
+	}
+
+	u := api.NewURL().Path(path...)
+
+	// Always set project parameter if provided (operations and warnings may be project specific but it is not a requirement).
+	if projectName != "" {
+		u = u.WithQuery("project", projectName)
+	}
+
+	// Always set location if provided.
+	if location != "" {
+		u = u.WithQuery("target", location)
+	}
+
+	return Reference(u.String()), nil
```

Comment:
Now seeing this function, im unclear why we wouldnt return it as a *url.URL seeing as you've got one internally, and the function is called `URL()`?

Reply:
Maybe an `api.URL`. Then I can implement `json.Marshaler` and `json.Unmarshaler` for it as these will be the display format for permissions.

I will try it.

Reply:
Ok it's now returning a `*api.URL` :+1: